firebase-firestore/src/main/java/com/google/firebase/firestore/remote/GrpcCallProvider.java

```diff
@@ -77,7 +77,9 @@ public static void overrideChannelBuilder(
             () -> {
               ManagedChannel channel = initChannel(context, databaseInfo);
               FirestoreGrpc.FirestoreStub firestoreStub =
-                  FirestoreGrpc.newStub(channel).withCallCredentials(firestoreHeaders);
+                  FirestoreGrpc.newStub(channel)
+                      .withCallCredentials(firestoreHeaders)
+                      .withExecutor(asyncQueue.getExecutor());
```

Comment:
I don't understand why this isn't a noop? It seems like we already set the executor down on line 122, and my understanding of https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/AbstractStub.html#withExecutor-java.util.concurrent.Executor- would be that the executor set on 122 should apply here.

Reply:
The debug view when I hit the codepath from the PR description:

<img width="1406" alt="Screen Shot 2019-07-08 at 1 38 55 PM" src="https://user-images.githubusercontent.com/19397744/60841261-3cd82a80-a186-11e9-9f19-dbf8198359c3.png">

The executor here is set to the our queue, but only because we don't have an override via CallOptions. It seems relying on a default can cause weirdness if two Firestore clients are active at the same time. If we specify the executor on a per-call level, this weirdness goes away (I have updated the PR to no longer set the executor on the channel itself).

I wish I fully understood what is going on here, but I can't seem to get to the bottom of this. I do however have a 100% failure rate without this PR (for a full run of all integration tests against the emulator) and a 100% success rate with it.

If you have any further insight that could help me debug this, let me know :(

Reply:
Oh, I think I see it. Maybe.

We build up the stub, but then immediately discard it (first saving the CallOptions.) The CallOptions only saves the options set on the stub, and doesn't take into account the options set on the channel within the stub. (The stub passes it's CallOptions to it's channel when making an rpc.)

Then, we call ManagedChannel::newCall, passing in the calloptions. (The channel itself already has the correct executor, so leaving it null in calloptions *should* work, though apparently doesn't.)

Eventually, we create a https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ClientCallImpl.java, passing in those same CallOptions. Then we call start() on that.

We eventually create the ClientTransport (which is presumably the DelayedClientTransport one that you linked) here: https://github.com/grpc/grpc-java/blob/cc13f74065911dab19c4e757cf559fd84ff3207e/core/src/main/java/io/grpc/internal/ClientCallImpl.java#L257. Note that it gets passed the callOptions, but not the executor.

I don't know if that's a bug in grpc, or is WAI. But while I'm still murky on the details, I think it does explain why your fix works. I think this is worth involving the grpc people to confirm one way or the other.

But for now, your fix lgtm.

Reply:
I am not sure if I know have even more questions :) I was never able to fully figure out all the interactions. I will file an issue in the GRPC-Java repo to figure out what the root cause is. Thank you!

Reply:
Filed https://github.com/grpc/grpc-java/issues/5966